src/modules/mc_pos_control/MulticopterPositionControl.cpp

```diff
@@ -568,14 +568,29 @@ void MulticopterPositionControl::Run()
 
 			_control.setState(states);
 
+			const hrt_abstime now = hrt_absolute_time();
+
 			// Run position control
-			if (!_control.update(dt)) {
-				// Failsafe
-				_vehicle_constraints = {0, NAN, NAN, false, {}}; // reset constraints
+			if (_control.update(dt)) {
+
+				// Valid control update - store for fallback
+				_last_valid_setpoint_time = now;
+				_last_valid_setpoint = _setpoint;
+
+			} else {
 
-				_control.setInputSetpoint(generateFailsafeSetpoint(vehicle_local_position.timestamp_sample, states, true));
-				_control.setVelocityLimits(_param_mpc_xy_vel_max.get(), _param_mpc_z_vel_max_up.get(), _param_mpc_z_vel_max_dn.get());
-				_control.update(dt);
+				if ((now - _last_valid_setpoint_time) < 100_ms) {
+					// Use last valid setpoint for the timeout period
+					_control.setInputSetpoint(_last_valid_setpoint);
+
+				} else {
+					// Failsafe
+					_vehicle_constraints = {0, NAN, NAN, false, {}}; // reset constraints
+
+					_control.setInputSetpoint(generateFailsafeSetpoint(vehicle_local_position.timestamp_sample, states, true));
+					_control.setVelocityLimits(_param_mpc_xy_vel_max.get(), _param_mpc_z_vel_max_up.get(), _param_mpc_z_vel_max_dn.get());
+					_control.update(dt);
```

Comment:
Shouldn't this be done after the else because the controller needs to run with either _last_valid_setpoint or failsafe setpoint?